query/src/main/scala/filodb/query/exec/ApplySubqueryRangeFunction.scala

```diff
@@ -0,0 +1,155 @@
+          subqueryExpRangeParams.stepSecs * 1000,
+          underlyingLookbackMs,
+          underlyingStepMs,
+          rv.rows,
+          rangeFuncGen,
+          querySession.queryConfig
+        ),
+        None
+      )
+    }
+  }
+
+  override def schema(source: ResultSchema): ResultSchema = {
+    // since range function is NOT optional, the resulting vectors length will be
+    // of size 1, not the length of the underlying PeriodicSeries
+    source.copy(fixedVectorLen = Some(1))
+  }
+
+}
+
+// sum_over_time(foo[5m:1m])[3m:90s]
+class SubquerySlidingWindowIterator(
+   startMs: Long, // now - 3m
+   endMs : Long, // now
+   stepMs: Long, // 90s
+   underlyingLookbackMs: Long, //5m
+   underlyingStepMs: Long, //1m
+   rvc: RangeVectorCursor,
+   rfGenerator: RangeFunctionGenerator, //TODO require reset for all range functions
+   //rangeFunction: RangeFunction,
+   queryConfig: QueryConfig
+ ) extends RangeVectorCursor {
+  private val sampleToEmit = new TransientRow()
+
+  // samples queue that will have ALL of the samples in the current range
+  // whether they are applicable for the current window or not
+  private val rangeQueue = new IndexedArrayQueue[TransientRow]()
+
+  // to avoid creation of object per sample, we use a pool
+  val windowSamplesPool = new TransientRowPool()
+
+  private var ts = 0L
+  private var curWindowStartMs = startMs - underlyingLookbackMs
+  private var curWindowEndMs = startMs
+  private var finished = false
+
+  override def close(): Unit = rvc.close()
+
+  override def hasNext: Boolean = {
+    if (finished) {
+      false
+    } else {
+      rvc.hasNext
+    }
+  }
+
+  override def next(): TransientRow = {
+    cleanBuffer()
+    addSamplesToBuffer()
+    val rangeFunction : RangeFunction = rfGenerator().asSliding
+    // samples queue
+    val windowQueue = new IndexedArrayQueue[TransientRow]()
+    // this is the object that will be exposed to the RangeFunction
+    val windowSamples = new QueueBasedWindow(windowQueue)
+    for (i <- 0 until rangeQueue.size) {
+      val sample = rangeQueue.apply(i)
+      // not ALL underlying periodic series samples need to be used for calculation
+      // of the results of this particular window
+      if (
+        sample.timestamp <= curWindowEndMs &&
+        (sample.timestamp - curWindowStartMs) % underlyingStepMs == 0
+      ) {
+        windowQueue.add(sample)
+        rangeFunction.addedToWindow(sample, windowSamples)
+      }
+    }
+    rangeFunction.apply(curWindowStartMs, curWindowEndMs, windowSamples, sampleToEmit, queryConfig)
+    curWindowStartMs += stepMs
+    curWindowEndMs += stepMs
+    sampleToEmit
+  }
+
+  def cleanBuffer() : Unit = {
+    while (rangeQueue.size > 0 && rangeQueue.head.timestamp < curWindowStartMs) {
+      rangeQueue.remove()
+    }
+  }
+
+  def addSamplesToBuffer() : Unit = {
+    while(ts < curWindowEndMs && ts <=endMs && rvc.hasNext) {
+      val cur = rvc.next()
+
+      val toAdd = windowSamplesPool.get
```

Comment:
I see a get, but no call to return the object back to pool.
Can we please confirm that there is no memory accumulation as we proceed with the query execution?

Reply:
In case this class is invoked I will explicitly throw an exception (experimental etc etc, so, the developer who might be interested can actually modify it)

Reply:
removed an entire class